// @remotion/eslint-plugin/dist/rules/v4-import.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("@typescript-eslint/utils");
const createRule = utils_1.ESLintUtils.RuleCreator(() => {
    return `https://www.remotion.dev/docs/4-0-migration`;
});
const ImportConfig = "Update the import to the new V4 location: import {Config} from '@remotion/cli/config'";
const rule = createRule({
    name: "v4-config-import",
    meta: {
        type: "problem",
        docs: {
            description: ImportConfig,
            recommended: "warn",
        },
        fixable: undefined,
        schema: [],
        messages: {
            ImportConfig,
        },
    },
    defaultOptions: [],
    create: (context) => {
        return {
            ImportDeclaration: (node) => {
                if (node.source.value !== "remotion") {
                    return;
                }
                const config = node.specifiers.find((s) => s.type === "ImportSpecifier" && s.imported.name === "Config");
                if (config) {
                    context.report({
                        messageId: "ImportConfig",
                        node,
                    });
                }
            },
        };
    },
});
exports.default = rule;
