// @remotion/cli/dist/config/chromium-flags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setChromiumMultiProcessOnLinux = exports.getChromiumMultiProcessOnLinux = exports.setChromiumHeadlessMode = exports.getChromiumHeadlessMode = exports.setChromiumOpenGlRenderer = exports.getChromiumOpenGlRenderer = exports.setChromiumIgnoreCertificateErrors = exports.getIgnoreCertificateErrors = exports.setChromiumDisableWebSecurity = exports.getChromiumDisableWebSecurity = void 0;
const renderer_1 = require("@remotion/renderer");
let chromiumDisableWebSecurity = false;
let ignoreCertificateErrors = false;
let openGlRenderer = renderer_1.RenderInternals.DEFAULT_OPENGL_RENDERER;
let headlessMode = true;
let multiProcessOnLinux = false;
const getChromiumDisableWebSecurity = () => chromiumDisableWebSecurity;
exports.getChromiumDisableWebSecurity = getChromiumDisableWebSecurity;
const setChromiumDisableWebSecurity = (should) => {
    chromiumDisableWebSecurity = should;
};
exports.setChromiumDisableWebSecurity = setChromiumDisableWebSecurity;
const getIgnoreCertificateErrors = () => ignoreCertificateErrors;
exports.getIgnoreCertificateErrors = getIgnoreCertificateErrors;
const setChromiumIgnoreCertificateErrors = (should) => {
    ignoreCertificateErrors = should;
};
exports.setChromiumIgnoreCertificateErrors = setChromiumIgnoreCertificateErrors;
const getChromiumOpenGlRenderer = () => openGlRenderer;
exports.getChromiumOpenGlRenderer = getChromiumOpenGlRenderer;
const setChromiumOpenGlRenderer = (renderer) => {
    renderer_1.RenderInternals.validateOpenGlRenderer(renderer);
    openGlRenderer = renderer;
};
exports.setChromiumOpenGlRenderer = setChromiumOpenGlRenderer;
const getChromiumHeadlessMode = () => headlessMode;
exports.getChromiumHeadlessMode = getChromiumHeadlessMode;
const setChromiumHeadlessMode = (should) => {
    headlessMode = should;
};
exports.setChromiumHeadlessMode = setChromiumHeadlessMode;
const getChromiumMultiProcessOnLinux = () => multiProcessOnLinux;
exports.getChromiumMultiProcessOnLinux = getChromiumMultiProcessOnLinux;
const setChromiumMultiProcessOnLinux = (should) => {
    multiProcessOnLinux = should;
};
exports.setChromiumMultiProcessOnLinux = setChromiumMultiProcessOnLinux;
