// @remotion/renderer/dist/options/color-space.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateColorSpace = exports.colorSpaceOption = exports.validColorSpaces = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
exports.validColorSpaces = ['default', 'bt709'];
exports.colorSpaceOption = {
    name: 'Color space',
    cliFlag: 'color-space',
    description: () => ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: ["Color space to use for the video. Acceptable values:", ' ', (0, jsx_runtime_1.jsxs)("code", { children: ['"', "default", '"'] }), ",", ' ', (0, jsx_runtime_1.jsxs)("code", { children: ['"', "bt709", '"'] })] })),
    docLink: 'https://www.remotion.dev/docs/renderer/render-media#colorspace',
    ssrName: 'colorSpace',
    type: 'default',
};
const validateColorSpace = (option) => {
    if (exports.validColorSpaces.includes(option)) {
        return;
    }
    throw new Error(`Expected one of ${exports.validColorSpaces.map((c) => `"${c}"`).join(', ')} for ${exports.colorSpaceOption.ssrName} but got "${option}"`);
};
exports.validateColorSpace = validateColorSpace;
