// @remotion/renderer/dist/options/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.allOptions = void 0;
const audio_bitrate_1 = require("./audio-bitrate");
const color_space_1 = require("./color-space");
const crf_1 = require("./crf");
const delete_after_1 = require("./delete-after");
const enable_multiprocess_on_linux_1 = require("./enable-multiprocess-on-linux");
const enforce_audio_1 = require("./enforce-audio");
const folder_expiry_1 = require("./folder-expiry");
const gl_1 = require("./gl");
const jpeg_quality_1 = require("./jpeg-quality");
const mute_1 = require("./mute");
const offthreadvideo_cache_size_1 = require("./offthreadvideo-cache-size");
const scale_1 = require("./scale");
const video_bitrate_1 = require("./video-bitrate");
const video_codec_1 = require("./video-codec");
const webhook_custom_data_1 = require("./webhook-custom-data");
exports.allOptions = {
    scaleOption: scale_1.scaleOption,
    crfOption: crf_1.crfOption,
    jpegQualityOption: jpeg_quality_1.jpegQualityOption,
    videoBitrate: video_bitrate_1.videoBitrate,
    audioBitrateOption: audio_bitrate_1.audioBitrateOption,
    enforceAudioOption: enforce_audio_1.enforceAudioOption,
    muteOption: mute_1.muteOption,
    videoCodecOption: video_codec_1.videoCodecOption,
    offthreadVideoCacheSizeInBytesOption: offthreadvideo_cache_size_1.offthreadVideoCacheSizeInBytesOption,
    webhookCustomDataOption: webhook_custom_data_1.webhookCustomDataOption,
    colorSpaceOption: color_space_1.colorSpaceOption,
    deleteAfterOption: delete_after_1.deleteAfterOption,
    folderExpiryOption: folder_expiry_1.folderExpiryOption,
    enableMultiprocessOnLinuxOption: enable_multiprocess_on_linux_1.enableMultiprocessOnLinuxOption,
    glOption: gl_1.glOption,
};
