// @remotion/renderer/dist/options/options-map.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.optionsMap = void 0;
const offthreadvideo_cache_size_1 = require("./offthreadvideo-cache-size");
exports.optionsMap = {
    renderMedia: [offthreadvideo_cache_size_1.offthreadVideoCacheSizeInBytesOption],
    renderStill: [offthreadvideo_cache_size_1.offthreadVideoCacheSizeInBytesOption],
    getCompositions: [offthreadvideo_cache_size_1.offthreadVideoCacheSizeInBytesOption],
    selectComposition: [offthreadvideo_cache_size_1.offthreadVideoCacheSizeInBytesOption],
    renderFrames: [offthreadvideo_cache_size_1.offthreadVideoCacheSizeInBytesOption],
    renderMediaOnLambda: [offthreadvideo_cache_size_1.offthreadVideoCacheSizeInBytesOption],
    renderStillOnLambda: [offthreadvideo_cache_size_1.offthreadVideoCacheSizeInBytesOption],
    getCompositionsOnLambda: [offthreadvideo_cache_size_1.offthreadVideoCacheSizeInBytesOption],
    renderMediaOnCloudRun: [offthreadvideo_cache_size_1.offthreadVideoCacheSizeInBytesOption],
    renderStillOnCloudRun: [offthreadvideo_cache_size_1.offthreadVideoCacheSizeInBytesOption],
};
